execution/executor-types/src/in_memory_state_calculator.rs

```diff
@@ -124,63 +123,46 @@ impl InMemoryStateCalculator {
         txn: &Transaction,
         txn_output: &ParsedTransactionOutput,
     ) -> Result<(HashMap<StateKey, StateValue>, Option<HashValue>)> {
-        let updated_state_keys = process_write_set(
+        let updated_state_kvs = process_write_set(
             Some(txn),
             &mut self.state_cache,
             txn_output.write_set().clone(),
         )?;
-        self.updated_after_latest
-            .extend(updated_state_keys.into_iter());
+        self.updates_after_latest.extend(updated_state_kvs.clone());
         self.next_version += 1;
 
         if txn_output.is_reconfig() {
-            self.checkpoint()
+            Ok((updated_state_kvs, Some(self.checkpoint_root_hash()?)))
         } else {
             match txn {
                 Transaction::BlockMetadata(_) | Transaction::UserTransaction(_) => {
-                    Ok((HashMap::new(), None))
+                    Ok((updated_state_kvs, None))
                 }
                 Transaction::GenesisTransaction(_) | Transaction::StateCheckpoint(_) => {
-                    self.checkpoint()
+                    Ok((updated_state_kvs, Some(self.checkpoint_root_hash()?)))
                 }
             }
         }
     }
 
-    fn checkpoint(&mut self) -> Result<(HashMap<StateKey, StateValue>, Option<HashValue>)> {
+    fn checkpoint_root_hash(&mut self) -> Result<HashValue> {
```

Comment:
nit: naming: the original name `checkpoint()` was meant to be a verb.. so how about calling it `make_checkpoint()` if you don't like the short name? :P